{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "76c1d02e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Transalte the dna Sequence into Protein\n",
    "# DNA -> complement -> mRNA -> codons -> amino acids\n",
    "# DNA -> Complemnt\n",
    "# A -> T\n",
    "# T -> A\n",
    "# G -> C\n",
    "# C -> G\n",
    "\n",
    "# Complemnt -> mRNA\n",
    "# T -> U\n",
    "\n",
    "# mRNA -> codons\n",
    "# Seperate to 3 \n",
    "\n",
    "# codons -> amino acids\n",
    "# Codon Table\n",
    "\n",
    "# DNA:  ATG TCC TTT TTA TCG\n",
    "# mRNA: UAC AGG AAA AAU AGC\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "10ed6177",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original DNA:        ATGTCCTTTTTATCG\n",
      "Complementary DNA:   TACAGGAAAAATAGC\n",
      "mRNA Sequence:       UACAGGAAAAAUAGC\n"
     ]
    }
   ],
   "source": [
    "# dnaSeq=['A','T','G','T','C','C','T','T','T','T','T','A','T','C','G']\n",
    "dnaSeq='ATGTCCTTTTTATCG'\n",
    "def convert_Complemnt(dnaSeq):\n",
    "    complement=\"\"\n",
    "    for base in dnaSeq:\n",
    "        if base == 'A':\n",
    "            complement += \"T\"\n",
    "        elif base == 'T':\n",
    "            complement += \"A\"\n",
    "        elif base == 'G':\n",
    "            complement += \"C\"\n",
    "        elif base == 'C':\n",
    "            complement += \"G\"\n",
    "    return complement\n",
    "\n",
    "def convert_mRNA(dnaSeq):\n",
    "    mRNA=\"\"\n",
    "    for base in complement:\n",
    "        if base == 'A':\n",
    "            mRNA += \"A\"\n",
    "        elif base == 'T':\n",
    "            mRNA += \"U\"\n",
    "        elif base =='G':\n",
    "            mRNA += \"G\"\n",
    "        elif base == 'C':\n",
    "            mRNA += \"C\"\n",
    "    return mRNA\n",
    "        \n",
    "\n",
    "complement= convert_Complemnt(dnaSeq)\n",
    "mRNA= convert_mRNA(complement)\n",
    "print(\"Original DNA:       \", dnaSeq)\n",
    "print(\"Complementary DNA:  \", complement)\n",
    "print(\"mRNA Sequence:      \", mRNA)\n",
    "        \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "cc6c0bc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Codon Sequence:  UAC-AGG-AAA-AAU-AGC\n",
      "Protein Sequence: Tyr-Arg-Lys-Asn-Ser\n"
     ]
    }
   ],
   "source": [
    "def convert_Protein(dnaSeq):\n",
    "    protein = \"\"\n",
    "    codon_list=[]\n",
    "    for i in range(0, len(mRNA), 3):\n",
    "        codon = mRNA[i:i+3]\n",
    "        codon_list.append(codon)\n",
    "\n",
    "        if codon == 'UAC':\n",
    "            protein += \"Tyr-\"\n",
    "        elif codon == 'AGG':\n",
    "            protein += \"Arg-\"\n",
    "        elif codon == 'AAA':\n",
    "            protein += \"Lys-\"\n",
    "        elif codon == 'AAU':\n",
    "            protein += \"Asn-\"\n",
    "        elif codon == 'AGC':\n",
    "            protein += \"Ser-\"\n",
    "        elif codon == 'AGG':\n",
    "            protein += \"Arg-\"\n",
    "\n",
    "\n",
    "    if protein.endswith(\"-\"):\n",
    "        protein = protein[:-1]\n",
    "    print(\"Codon Sequence: \", '-'.join(codon_list))\n",
    "    return protein\n",
    "\n",
    "\n",
    "protein= convert_Protein(mRNA)\n",
    "print(\"Protein Sequence:\", protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6714140d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mutated DNA Seq:  ATGTCATTTTTATCG\n",
      "Codon Sequence:  UAC-AGG-AAA-AAU-AGC\n",
      "Complementary Mutated DNA:  TACAGTAAAAATAGC\n",
      "Mutated mRNA Sequence:  UACAGGAAAAAUAGC\n",
      "Mutated Protein Sequence: Tyr-Arg-Lys-Asn-Ser\n"
     ]
    }
   ],
   "source": [
    "def point_mutation(dnaSeq, position, new_base):\n",
    "    if position < 0 or position >= len(dnaSeq):\n",
    "        return \"Error: Position out of range.\"\n",
    "    if new_base not in 'ATGC':\n",
    "        return \"Error: Invalid nucleotide.\"\n",
    "    \n",
    "    dna_list = list(dnaSeq)\n",
    "    dna_list[position] = new_base \n",
    "    mutated_seq = ''.join(dna_list)\n",
    "    return mutated_seq\n",
    "\n",
    "mutated = point_mutation(dnaSeq, 5, 'A')\n",
    "print(\"Mutated DNA Seq: \", mutated)\n",
    "\n",
    "complement_mutated= convert_Complemnt(mutated)\n",
    "mRNA_mutated= convert_mRNA(complement_mutated)\n",
    "mutated_protein= convert_Protein(mRNA_mutated)\n",
    "print(\"Complementary Mutated DNA: \", complement_mutated)\n",
    "print(\"Mutated mRNA Sequence: \", mRNA_mutated)\n",
    "print(\"Mutated Protein Sequence:\", mutated_protein)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
