{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dd33a24d-7f5f-41d0-84bb-d1a44a948b41",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = \"AGCTTTCGAATTCGA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5a1d805-3293-491e-956d-c2b601f566ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Find all substrings\n",
    "subtrings=[]\n",
    "\n",
    "for i in range( len(dna)):\n",
    "    for j in range(i+1, len(dna)+1):\n",
    "        subtrings.append(dna[i:j])\n",
    "\n",
    "subtrings\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "43e3fda6-d94a-4473-98a7-1286306f7ba7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A', 'G', 'C', 'T', 'TT', 'TTC', 'TTCG', 'TTCGA', 'TC', 'TCG', 'TCGA', 'CG', 'CGA', 'GA']\n"
     ]
    }
   ],
   "source": [
    "#Find repeated substrings\n",
    "repeated = []\n",
    "\n",
    "for i in range(len(dna)):\n",
    "    for j in range(i+1, len(dna)+1):\n",
    "        sub = dna[i:j]\n",
    "        if dna.count(sub) > 1 and sub not in repeated:\n",
    "            repeated.append(sub)\n",
    "\n",
    "print(repeated)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dec87d17-8449-44fc-8517-2dd5e4d1b6f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A: 4 T: 5 G: 3 C: 3\n"
     ]
    }
   ],
   "source": [
    "#Count nucleotide frequency\n",
    "a = dna.count(\"A\")\n",
    "t = dna.count(\"T\")\n",
    "g = dna.count(\"G\")\n",
    "c = dna.count(\"C\")\n",
    "\n",
    "print(\"A:\", a, \"T:\", t, \"G:\", g, \"C:\", c)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "99b41125-31fc-4d28-8fd8-376a146a7e7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCTTTCGAATTCGA\n"
     ]
    }
   ],
   "source": [
    "#Remove invalid characters\n",
    "clean = \"\"\n",
    "\n",
    "for ch in dna:\n",
    "    if ch in \"ATGC\":\n",
    "        clean += ch\n",
    "\n",
    "print(clean)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "979a666b-0813-467c-9a76-20e3c359e250",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Not Palindrome\n"
     ]
    }
   ],
   "source": [
    "#Check if DNA is palindrome\n",
    "if dna == dna[::-1]:\n",
    "    print(\"Palindrome\")\n",
    "else:\n",
    "    print(\"Not Palindrome\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6c18c95c-e42c-40ba-b202-007d30a05ee6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AGCTTTCGA', 'GCTTTCG', 'CTTTC']\n"
     ]
    }
   ],
   "source": [
    "#Palindromes longer than 3\n",
    "palindromes = []\n",
    "\n",
    "for i in range(len(dna)):\n",
    "    for j in range(i+1, len(dna)+1):\n",
    "        sub = dna[i:j]\n",
    "        if sub == sub[::-1] and len(sub) > 3:\n",
    "            if sub not in palindromes:\n",
    "                palindromes.append(sub)\n",
    "\n",
    "print(palindromes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d5fbe275-f8e1-43d4-81de-94bbeb699de7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23\n"
     ]
    }
   ],
   "source": [
    "#Count palindromes\n",
    "count = 0\n",
    "\n",
    "for i in range(len(dna)):\n",
    "    for j in range(i+1, len(dna)+1):\n",
    "        sub = dna[i:j]\n",
    "        if sub == sub[::-1]:\n",
    "            count += 1\n",
    "\n",
    "print(count)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d51736da-b0bc-4688-a1d5-18df1b9a5274",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found at 7\n"
     ]
    }
   ],
   "source": [
    "#Find motif occurrences\n",
    "motif = \"GAATTC\"\n",
    "\n",
    "for i in range(len(dna)-len(motif)+1):\n",
    "    if dna[i:i+len(motif)] == motif:\n",
    "        print(\"Found at\", i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4331404b-c9a4-4ddc-a9ed-46ae598303a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "#Count motif\n",
    "count = 0\n",
    "\n",
    "for i in range(len(dna)-len(motif)+1):\n",
    "    if dna[i:i+len(motif)] == motif:\n",
    "        count += 1\n",
    "\n",
    "print(count)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5a038d2c-0640-4421-9294-e2333a1bf982",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find start codon\n",
    "for i in range(0, len(dna)-2):\n",
    "    if dna[i:i+3] == \"ATG\":\n",
    "        print(\"Start codon at\", i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "52ddeae3-8f06-406d-8bf6-e04b265ced6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find stop codons\n",
    "stops = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "\n",
    "for i in range(len(dna)-2):\n",
    "    if dna[i:i+3] in stops:\n",
    "        print(\"Stop codon at\", i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dae011c0-b893-451e-864a-ad42c85efd16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CTTTC', 'GCTTTCG', 'AGCTTTCGA']\n"
     ]
    }
   ],
   "source": [
    "#Sort palindromes by length\n",
    "palindromes.sort(key=len)\n",
    "print(palindromes)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
